source/Condition.h

```diff
@@ -0,0 +1,194 @@
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE. See the GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License along with
+this program. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#ifndef RVALUE_H_
+#define RVALUE_H_
+
+
+#include <cmath>
+#include <limits>
+#include <string>
+#include <type_traits>
+
+
+
+// This class stores either:
+//   1. A condition's value and key
+//   2. A literal value (key is empty)
+//
+// The value (V) is first in the template since (nearly?) all Condition
+// classes will use a std::string key (K). Having a template key type
+// allows storage of scope, wstring, etc. in the future without
+// rewriting this class.
+template < class V, class K = std::string >
+class Condition {
+public:
+	typedef V ValueType;
+	typedef K KeyType;
+
+	constexpr Condition() : value(), key() {}
+	explicit constexpr Condition(const V &value) : value(value), key() {}
+	constexpr Condition(const V &value, const K &key) : value(value), key(key) {}
+	~Condition() {}
+
+	template <class V2, class K2>
+	Condition(const Condition<V2,K2> &other);
+
+	template <class V2, class K2>
+	Condition<V,K> &operator = (const Condition<V2,K2> &other);
+
+	// Update the value from a scope that contains it
+	template<class Getter>
+	const ValueType &UpdateConditions(const Getter &getter);
+
+	// Update the value from a scope that contains it, but use
+	// ValueType() if Validator(weight) is false.
+	template<class Getter, class Validator>
+	const ValueType &UpdateConditions(const Getter &getter, Validator validator);
+
+	// Accessors and mutators
+
+	const ValueType &Value() const { return value; }
+	ValueType &Value() { return value; }
+	const KeyType &Key() const { return key; }
+	KeyType &Key() { return key; }
+
+	// Does this Condition come from the same place as the other one?
+	// If it was a condition, the key must be the same (value doesn't matter)
+	// If it was a literal (no key) then the value must be the same.
+	// If one is literal and the other is conditional, the result is false.
+	bool SameOrigin(const Condition<V,K> &o);
+
+	// Does this originate from a condition?
+	bool HasConditions() const { return !key.empty(); }
+
+	// Does this originate from a literal value (ie. 5.071)?
+	bool IsLiteral() const { return key.empty(); }
+
+	// Floating-point values are false if they're within half the
+	// type's precision of 0 while any other types are passed
+	// through static_cast<bool>
+	explicit operator bool() const;
+
+	// Allow the Condition to be treated as its value in most contexts.
+	operator ValueType() const { return value; }
+
+private:
+	ValueType value;
+	KeyType key;
```

Comment:
just thinking, maybe rename this to make it clearer? smth like referenceKey?

also does it really have to be a template too?

Reply:
Hard-coded types in a template is generally considered a bad practice, since the purpose of a template is to increase abstraction. There is no cost to making it a template parameter. So, no, it doesn't *need* to be a template parameter, but it *should* be a template parameter.

Reply:
similar as for the other one I'm asking what kind of things it could hold

Reply:
It has to a valid key type in the Getter.HasGet in UpdateConditions, and it needs an empty() method to determine if it is empty. I'm not sure how to require an empty() method. I know there's a way to put that in a static assertion. Will investigate...

Reply:
There is now a static assertion for a KeyType::empty.

Reply:
well thats not what I meant, I meant what kind of keys are we giving it?

Reply:
The code never requests a key type, so it uses the default: std::string. That's why the second template argument is never mentioned in the PR description.

Reply:
Did you mean a comment in the code in Condition.h? No, I'd rather the comments in Condition be specific to the functionality in that file. The Condition should not need to know about how it is used elsewhere. It should, however, be able to restrict how it is used, which is why it now has static assertions. If there was an Endless Sky design document, or similar, then such a comment would go there.

Reply:
I think I came up with a compromise. There is a long comment at the top of `class Condition` explaining these things without explicitly stating what the rest of the code does.

In the test_condition.cpp unit tests, there is an "implementation note" that says the std::string is the default and the code assumes that. There are unit tests that check if Condition's default KeyType acts how it should.

Reply:
does it ever get smth else than string though?
do we really need the template *here* is what I'm asking basically caus if its always a string then just make it a string

Reply:
No, but I am of the opposite opinion. If a template class does not need a hard-coded type, it should not have one.